### ISSUE TITLE
Blank SurfaceView occasionally

### ISSUE THREAD

**Issue by JaseElder** [google/ExoPlayer#1748]
Hi,

Building an interactive storybook type app, with a full screen video on each "page". The pages are fragments that each contain a SurfaceView and are replaced using a FragmentManager replace() transaction.
All videos display and play correctly, but not all the time. On occasion, when paging through, a video will appear blank, as though it's not loaded into the SurfaceView. Paging out, and then back will cause the video to appear. Also minimising and resuming.
The bug doesn't happen on a particular video - any of them could show or not show.

Running on Nexus 9 with Android 6.0.1.

*Edit: Also, this was with ExoPlayer 1.5.9 and now 1.5.10

I tried replicating the DemoPlayer wrapper code, complete with SurfaceView lifecycle handling, and the same thing happens.
I've tested on a Huawei P8 with Android 6.0 but haven't been able to reproduce the issue on that device.

I've noticed in the logs that when the issue **doesn't** occur, there are two SurfaceUtils **"set up nativeWindow"** messages as in the below:

```
08-16 19:58:21.922 E/BaseFragment: surfaceCreated: BedroomPageFragment
08-16 19:58:21.946 I/OMXClient: Using client-side OMX mux.
08-16 19:58:22.033 I/OMXClient: Using client-side OMX mux.
08-16 19:58:22.039 I/MediaCodec: [OMX.Nvidia.h264.decode] setting surface generation to 460820
08-16 19:58:22.052 D/SurfaceUtils: set up nativeWindow 0x558cdbc080 for 1024x768, color 0x106, rotation 0, usage 0x2b00
08-16 19:58:22.086 D/SurfaceUtils: set up nativeWindow 0x558cdbc080 for 1024x768, color 0x106, rotation 0, usage 0x2b00
08-16 19:58:22.087 W/ACodec: [OMX.Nvidia.h264.decode] setting nBufferCountActual to 21 failed: -1010
08-16 19:58:22.087 W/ACodec: [OMX.Nvidia.h264.decode] setting nBufferCountActual to 20 failed: -1010
08-16 19:58:22.151 E/BaseFragment: surfaceDestroyed: EndPageFragment
08-16 19:58:22.151 W/ExoPlayerImplInternal: Sent message(1) after release. Message ignored.
08-16 19:58:22.177 E/BaseFragment: onDrawnToSurface: drawn
```

(also notice the MediaCodecVideoTrackRenderer.EventListener onDrawnToSurface listener being called)

But when the issue **does** occur, there is only one SurfaceUtils "set up nativeWindow" message, as in the below:

```
08-16 20:44:19.791  E/BaseFragment: surfaceCreated: BlindsPageFragment
08-16 20:44:19.808 I/OMXClient: Using client-side OMX mux.
08-16 20:44:19.833 I/OMXClient: Using client-side OMX mux.
08-16 20:44:19.836 I/MediaCodec: [OMX.Nvidia.h264.decode] setting surface generation to 20962307
08-16 20:44:19.845 D/SurfaceUtils: set up nativeWindow 0x558cd2d710 for 1024x768, color 0x106, rotation 0, usage 0x2b00
08-16 20:44:20.200 E/BaseFragment: surfaceDestroyed: LampshadePageFragment
08-16 20:44:20.200 W/ExoPlayerImplInternal: Sent message(1) after release. Message ignored.
```

and notice that the MediaCodecVideoTrackRenderer.EventListener onDrawnToSurface listener isn't called when the issue occurs.

The issue occurs whether paging quickly or slowly. 

Are the SurfaceUtils messages significant? I had a look at the source, but it appears to be part of a stagefright C++ file, so I'm not sure how to trace back from that to find out why nativeWindow isn't being set up a 2nd time (and also whether it should be)
https://android.googlesource.com/platform/frameworks/av/+/master/media/libstagefright/SurfaceUtils.cpp

Here's the full adbbugreport
[adbbugreport.txt.zip](https://github.com/google/ExoPlayer/files/421205/adbbugreport.txt.zip)

*Edit 2: Interestingly, if I put in a delay for sending the surface setting message to the player, it happens a lot less often.

```
    @Override
    public void surfaceCreated(final SurfaceHolder holder) {
        if (mPlayer != null) {
            Log.e(TAG, "surfaceCreated: " + getClass().getName());
            final Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    mPlayer.sendMessage(mVideoRenderer,
                            MediaCodecVideoTrackRenderer.MSG_SET_SURFACE, holder.getSurface());
                }
            }, 250);
        }
    }
```

**Comment by tprochazka** [comment]
I have the some problem after migrate to version 2.2.0.
It was working correctly on 2.0.2.
But I have another app, when 2.2.0 works correctly on the same stream.
Only difference I found is that "SurfaceUtils: set up nativeWindow" is displayed twice in the app when screen remain black. I not found any difference yet

**Comment by mufumbo** [comment]
same here. Happens to huwai P8 and samsung devices a lot. 
who can get this fixed without closing for not following the issue template?

**Comment by JaseElder** [comment]
Is there an issue template? Apologies for not following it.

**Comment by mufumbo** [comment]
Yes other similar issues have been closed because of that. Though you added a lot of information that makes it very replicable. Thanks so much. For now we implemented your thread sleep work around to avoid exo deadlock bugs.

**Comment by ojw28** [comment]
Is it possible to provide some (fairly minimal) patch to the ExoPlayer demo application that allows the issue to be reproduced there? If someone can get us that far then we'll take a look. Until then I don't think we have clear reproduction steps.

**Comment by mufumbo** [comment]
@ojw28 thanks for the follow up! 

You work at google? Can you please go to the device wardrobe and pick up a "Huawei P8" and try? From my users feedback, it happens 100% of the time in any video playback using ExoPlayer. 

We also have many users (30% of android) with "Samsung" manufacturer that are experiencing the same thing. 
I can't reproduce it either because I don't have a wardrobe full of devices like you do. Although, 30% of my users are on "Samsung" general brand. For *sure* it happens 100% of "Samsung S5", "Samsung S6" and "Samsung S7". 

Basically, there's no point in using ExoPlayer if we have to do this in our code:
```
return isLollipop() && !("samsung".equals(Build.MANUFACTURER));
```

It's very unfortunate because we might have to remove ExoPlayer to simplify our code since we had to implement a wrapper to support both VideoPlayer and ExoPlayer because of this type of bug.

**Comment by ojw28** [comment]
@mufumbo - ExoPlayer powers a *huge* number of playbacks every day across many apps on Samsung devices, including millions of YouTube playbacks, and we just don't see lots of users reporting that they see a blank screen instead of the video. We also test with the demo app regularly on Samsung devices and do not see this issue. That suggests there's something that your app(s) do specifically that causes the problem to reproduce, which is why we're requesting a patch to the demo app that also allows the issue to be reproduced there. This will help us to reproduce the issue and to understand the conditions necessary for it to occur.

If I had to guess, I suspect the demo app will work fine on a Huawei P8 as well, and that we'll need proper reproduction steps to debug on that device as well. But I'll try and track one down to give it a try (we do not have a wardrobe containing an unlimited number of every Android device ever shipped ;)).

**Comment by mufumbo** [comment]
thanks a lot @ojw28 !! 
ExoPlayer is the core piece for any android app that wants to run video streaming. 
Your hard work is very much appreciated.

**Comment by ojw28** [comment]
I got hold of a Huawei P8 Max running Marshmallow (we don't have the normal P8, or one running Lollipop). Playback appears to work perfectly in the demo app. So I think we're still stuck on needing proper reproduction steps as described above.

**Comment by mufumbo** [comment]
Thanks @ojw28 !! 

We just found that the bug was on our side indeed. We just need to set the TextureView properly in ExoPlayer. Our code was a bit complex since we use SurfaceView for the API level 24 and forward and weren't properly setting TextureView for API levels under 24. 

@JaseElder can you check if that's your problem too?

By the way, would it be wise for ExoPlayer demo to handle TextureView and SurfaceView rendering?

**Comment by JaseElder** [comment]
My issue is a bit different to the current - my code and the demo code works fine on the Huawei P8 (with ExoPlayer 1.5.10 anyway). It's only the Nexus 9 that I've experienced issues with so far. 
I'm on holiday on the other side of the planet at the moment, so won't be able to put something together for  about a month.

**Comment by ojw28** [comment]
Closing because one of the issues described here was determined to be an application issue, and we don't have reproduction steps for the other.
